spec-0007/transition_to_rng.py

```diff
@@ -0,0 +1,156 @@
+import numpy as np
+import functools
+import warnings
+
+
+def _transition_to_rng(old_name, position_num=None, dep_version=None):
+    """Example decorator to transition from old PRNG usage to new `rng` behavior
+
+    Suppose the decorator is applied to a function that used to accept parameter
+    `old_name='random_state'` either by keyword or as a positional argument at
+    `position_name=1`. At the time of application, the name of the argument in the
+    function signature is manually changed to the new name, `rng`. If positional
+    use was allowed before, this is not changed.*
+
+    - If the function is called with both `random_state` and `rng`, the decorator
+      raises an error.
+    - If `random_state` is provided as a keyword argument, the decorator passes
+      `random_state` to the function's `rng` argument as a keyword, but the decorator
+      will emit a `DeprecationWarning` about the deprecation of keyword `random_state`.
+    - If `random_state` is provided as a positional argument, the decorator passes
+      `random_state` to the function's `rng` argument by position, but the decorator
+      will emit a `FutureWarning` about the changing interpretation of the argument.
+    - If `rng` is provided as a keyword argument, the decorator validates `rng` using
+      `numpy.random.default_rng` before passing it to the function.
+    - If neither `random_state` nor `rng` is provided, the decorator checks whether
+      `np.random.seed` has been used to set the global seed. If so, it emits a
+      `FutureWarning`, noting that usage of `numpy.random.seed` will eventually have
+      no effect. Either way, the decorator calls the function without explicitly
+      passing the `rng` argument.
+
+    To avoid warnings, a user must pass `rng` as a keyword.
+
+    After the deprecation period, the decorator can be removed, and the function
+    can simply validate the `rng` argument by calling `np.random.default_rng(rng)`.
+
+    * A `FutureWarning` is emitted when the PRNG argument is used by
+      position. It indicates that the "Hinsen principle" (same
+      code yielding different results in two versions of the software)
+      will be violated, unless positional use is deprecated. Specifically:
+
+      - If `None` is passed by position, the function will change from being
```

Comment:
```suggestion
      - If `None` is passed by position and `np.random.seed`
        has been used, the function will change from being
```
For completeness, at least two other cases are problematic: if the user passes `np.random` or a `RandomState` by position, those will cause an error in the future. Not sure if these need to be mentioned in the text, but thought it was worth noting here.

Reply:
Addressed by https://github.com/stefanv/specs/commit/bd32f7b3377fb499e5e6da9b35db1e049debbc83.